SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m

```diff
@@ -290,18 +314,100 @@ - (nullable UIImage *)image
     }
 }
 
-- (nullable UIImage *)videoThumbnail
+- (nullable UIImage *)thumbnailImage
+{
+    NSString *thumbnailPath = self.thumbnailPath;
+    if (!thumbnailPath) {
+        OWSAssert(!self.isImage && !self.isVideo && !self.isAnimated);
+
+        return nil;
+    }
+
+    if (![[NSFileManager defaultManager] fileExistsAtPath:thumbnailPath]) {
+        OWSFail(@"%@ missing thumbnail for attachmentId: %@", self.logTag, self.uniqueId);
+
+        return nil;
+    }
+
+    return [UIImage imageWithContentsOfFile:self.thumbnailPath];
+}
+
+- (void)ensureThumbnail
+{
+    NSString *thumbnailPath = self.thumbnailPath;
+    if (!thumbnailPath) {
+        return;
+    }
+
+    if ([[NSFileManager defaultManager] fileExistsAtPath:thumbnailPath]) {
+        // already exists
+        return;
+    }
+
+    if (![[NSFileManager defaultManager] fileExistsAtPath:self.mediaURL.path]) {
+        OWSFail(@"%@ while generating thumbnail, source file doesn't exist: %@", self.logTag, self.mediaURL) return;
+    }
+
+    // TODO proper resolution?
+    CGFloat thumbnailSize = 200;
+
+    UIImage *_Nullable result;
+    if (self.isImage || self.isAnimated) {
+        CGImageSourceRef imageSource = CGImageSourceCreateWithURL((__bridge CFURLRef)self.mediaURL, NULL);
```

Comment:
FYI it's much faster to create thumbnails from a URL ImageSource vs. converting to/from PNG and using a Data ImageSource.